### PR TITLE
CTAA-1691_reminder_upload_keys_the_next_day

### PR DESCRIPTION
**Pull request by falkorichter**
## Changes

* remind user to upload the key from yesterday

Context: For privacy reasons, the key from today is not accessible until tomorrow.

Notification will link to this screen: 
<a href="https://user-images.githubusercontent.com/50506/85841772-e3a24600-b79e-11ea-8f14-1635f6a3734e.png"></a>

## Solved issues

- [Issue #1691](https://tasks.pxp-x.com/browse/CTAA-1691)

### CODE REVIEW

File: app/src/main/java/at/roteskreuz/stopcorona/App.kt [hunk 1]
```diff
@@ -33,6 +35,9 @@ class App : BaseApp() {
 
         // Create notification channels
         createNotificationChannels()
+
+        val notificationsRepository: NotificationsRepository = get()
+        notificationsRepository.checkAndScheduleMissingKeysNotification()
```
Comment (dusanjencik) [inline app/src/main/java/at/roteskreuz/stopcorona/App.kt]: Please move it to OfflineSyncer.

File: app/src/main/java/at/roteskreuz/stopcorona/model/repositories/NotificationsRepository.kt [hunk 2]
```diff
@@ -74,12 +77,25 @@ interface NotificationsRepository {
...
+    /**
+     * Display a notification, reminding the user to upload the keys from the day before.
+     * Context: For privacy reasons, the key from today is not accessible until tomorrow.
+     */
+    fun displayNotificationForUploadingKeysFromTheDayBefore()
+
+    /**
+     * observe
+     */
+    fun checkAndScheduleMissingKeysNotification()
```
Comment (dusanjencik) [inline app/src/main/java/at/roteskreuz/stopcorona/model/repositories/NotificationsRepository.kt]: I would suggest to move this method to `ExposureNotificationManager`.
Btw, update KDoc please.

File: app/src/main/java/at/roteskreuz/stopcorona/model/repositories/NotificationsRepository.kt [hunk 3]
```diff
@@ -221,6 +237,29 @@ class NotificationsRepositoryImpl(
...
+            pendingIntent = buildPendingIntentWithActivityStack {
+                addNextIntent(context.getDashboardActivityIntent().addFlags(firstActivityFlags))
+            },
+            channelId = NotificationChannels.CHANNEL_UPLOAD_KEYS,
+            ongoing = false
+        ).show()
+    }
+
+    override fun checkAndScheduleMissingKeysNotification(){
+        quarantineRepository.observeIfUploadOfMissingExposureKeysIsNeeded()
```
Comment (dusanjencik) [inline app/src/main/java/at/roteskreuz/stopcorona/model/repositories/NotificationsRepository.kt]: Cannot be this logic moved to setter of it? I guess we don't need to stay connected to this event when it's able to be started on demand.

File: app/src/main/java/at/roteskreuz/stopcorona/model/repositories/NotificationsRepository.kt [hunk 4]
```diff
@@ -221,6 +237,29 @@ class NotificationsRepositoryImpl(
...
+            .observeOnMainThread()
```
Comment (dusanjencik) [inline app/src/main/java/at/roteskreuz/stopcorona/model/repositories/NotificationsRepository.kt]: Please do not use Android framework in repositories if possible. Btw. Main thread here doesn't make any sense.

File: app/src/main/java/at/roteskreuz/stopcorona/model/workers/UploadKeysFromDayBeforeWorker.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    workerParams: WorkerParameters
+) : CoroutineWorker(appContext, workerParams), KoinComponent {
+
+    companion object {
+        private const val TAG = "UploadKeysFromDayBeforeWorker"
+
+
+        fun enqueueUploadKeysFromDayBeforeWorkerOnTheStartOfTheNextUtcDay(workManager: WorkManager) {
+            val request = OneTimeWorkRequestBuilder<UploadKeysFromDayBeforeWorker>()
+                .setInitialDelay(ZonedDateTime.now().millisUntilTheStartOfTheNextUtcDay(), TimeUnit.MILLISECONDS)
```
Comment (dusanjencik) [inline app/src/main/java/at/roteskreuz/stopcorona/model/workers/UploadKeysFromDayBeforeWorker.kt:31]: Do you really want to display a notification at midnight?

File: app/src/test/java/at/roteskreuz/stopcorona/model/entities/session/DbFullBatchPartTest.kt [hunk 6]
```diff
@@ -5,7 +5,7 @@ import org.junit.Test
 class DbFullBatchPartTest {
 
     @Test
-    fun getIntervalStart() {
+    fun testSomeSorting() {
```
Comment (dusanjencik) [inline app/src/test/java/at/roteskreuz/stopcorona/model/entities/session/DbFullBatchPartTest.kt]: Just informative: you can use
```
fun `descriptive long name of method often used in unit tests`()
```
Some is pretty vague. 😉

**Review by dusanjencik — CHANGES_REQUESTED** [review]

File: app/src/main/java/at/roteskreuz/stopcorona/model/workers/UploadKeysFromDayBeforeWorker.kt (hunk 5 above)
Comment (falkorichter) [reply app/src/main/java/at/roteskreuz/stopcorona/model/workers/UploadKeysFromDayBeforeWorker.kt:31]: Yes. This is what @uluckas decided. Show a notification in the middle of the night so people can see it when they wake up.

File: app/src/main/java/at/roteskreuz/stopcorona/App.kt (hunk 1 above)
Comment (dusanjencik) [reply app/src/main/java/at/roteskreuz/stopcorona/App.kt]: Deleted together.

File: app/src/main/java/at/roteskreuz/stopcorona/model/repositories/NotificationsRepository.kt (hunk 2 above)
Comment (dusanjencik) [reply app/src/main/java/at/roteskreuz/stopcorona/model/repositories/NotificationsRepository.kt]: Resolved together.

**Review by falkorichter — COMMENTED** [review]
@dusanjencik feedback as discussed pushed

File: app/src/main/java/at/roteskreuz/stopcorona/di/RepositoryModule.kt [hunk 7]
```diff
@@ -59,7 +59,9 @@ val repositoryModule = module {
         NotificationsRepositoryImpl(
             appDispatchers = get(),
             contextInteractor = get(),
-            dataPrivacyRepository = get()
+            dataPrivacyRepository = get(),
+            quarantineRepository = get(),
+            workManager = get()
```
Comment (dusanjencik) [inline app/src/main/java/at/roteskreuz/stopcorona/di/RepositoryModule.kt]: Not needed anymore.

**Comment by uliluckas** [comment]
Is it clear that we do not want to link directly to the key upload?

**Review by dusanjencik — DISMISSED** [review]
Nice 👍

File: app/src/main/java/at/roteskreuz/stopcorona/di/RepositoryModule.kt (hunk 7 above)
Comment (falkorichter) [reply app/src/main/java/at/roteskreuz/stopcorona/di/RepositoryModule.kt]: removed

**Review by dusanjencik — APPROVED** [review]

**Pull request merged**